### DISCUSSION TITLE
Get format as identified by ImageMagick

### DISCUSSION THREAD

**Discussion by aksdb** [libvips/libvips#3011] [General]
If I load a buffer through `magickload`, ImageMagick will perform its ... well ... magic, and determine how it can actually load it. Can I, in libvips, afterwards find out the format it determined? So if - for example - I load a `pbm` file, how would I know that it is `pbm`?

Use-case: I get images over the wire (potentially without correct filename) and want, after processing it using libvips, store it in the same format and also adjust the filename accordingly.

I assume the information is already there and I am just unable to find it in the docs. Any help is appreciated. Thanks :)

**Comment by jcupitt** [comment]
Hi @aksdb,

You can use eg. `vips_foreign_find_load_buffer()`:

https://www.libvips.org/API/current/VipsForeignSave.html#vips-foreign-find-load-buffer

And it'll return the name of the libvips load operator for that file format. There are similar things to sniff the format from a file, or from a generic data source, like a pipe.

You can also get the loader from the image metadata, eg. in python:

```
$ python3
Python 3.10.4 (main, Jun 29 2022, 12:14:53) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyvips
>>> x = pyvips.Image.new_from_file("k2.jpg")
>>> x.get("vips-loader")
'jpegload'
>>> 
```

That might be the best option for your use case. libvips images inherit metadata, so you could do this just before save.

```python
image = do_some_processing()
load_operation = image.get("vips-loader")
format = {"jpegload": "jpg", ...}
image.write_to_file(f"{filename}.{format}")
```

**Reply by jcupitt** [reply]
Oh, `vips-loader` won't be set if there's no source image. Maybe:

```python
if image.get_typeof("vips-load") == 0:
    format = "png" # default format
else:
    format = image.get("vips-loader")
```

**Reply by aksdb** [reply]
Hmm doesn't quite work.

```python
buf = open("example.pbm", "rb").read()
image = pyvips.Image.new_from_buffer(buf, "")
print(image.get("vips-loader"))
```

It prints "magickload_buffer". So I know it was loaded by imagemagick, but I don't know that it is a "pbm" file. ImageMagick apparently does know it (otherwise it couldn't load it, right?). So my question is: does ImageMagick expose that information to libvips, and does libvips expose it via API?

Maybe explained from another perspective: when I save a buffer using `magicksave`, I have to specify the `format` parameter. I basically look for the opposite. When loading something, I want to know the "format".

**Reply by jcupitt** [reply]
libvips has a nice PBM loader, perhaps it's been disabled in your build? I see:

```
$ vipsheader -f vips-loader marbles.pbm 
ppmload
```

I've added something to pick up the magick format in libvips, see https://github.com/libvips/libvips/commit/2e5e9fb7557266682fb4c3e2992bc655254e7755

**Reply by jcupitt** [reply]
... with that change I see:

```
$ vipsheader -f magick-format k2.bmp
BMP
```

**Reply by aksdb** [reply]
Awesome! That is exactly what I had in mind. Thank you very much!

**Reply by jcupitt** [reply]
I'd have a look at your libvips build as well -- it shouldn't be using imagemagick to load PBM.

**Reply by aksdb** [reply]
PBM just served as an example anyway. I am maintaining an image storage and transformation service, so I need to be able to handle as much formats as users can throw at it. While 99% is covered by the formats that libvips supports directly (jpeg, png, webp, heic, etc.), sometimes something more exotic gets uploaded and then it usually ends up with the ImageMagick fallback.

One of the use cases is that the image gets loaded, shrinked and then served in the original format. In that imagemagick-fallback-mode I had to rely on the file extension to issue the `save` command. But the file extension is not always correct and doesn't always match what imagemagick would call the `format`.

With your addition I can (in the future, once that is in a release and that release hits Alpine Linux :D) finally close that gap as well and keep using the format as expected by ImageMagick when saving.
